llvm/lib/SYCLLowerIR/LowerInvokeSimd.cpp

```diff
@@ -145,100 +155,327 @@ void getPossibleStoredVals(Value *Addr, ValueSetImpl &Vals) {
-        getPossibleStoredVals(LI->getPointerOperand(), Vals);
+        if (!collectPossibleStoredVals(LI->getPointerOperand(), Vals)) {
+          return false;
+        }
       } else {
         Vals.insert(V);
       }
       continue;
     }
     if (const auto *CI = dyn_cast<CallInst>(V)) {
       // only __builtin_invoke_simd is allowed, otherwise the pointer escapes
-      if (!getInvokeeIfInvokeSimdCall(CI)) {
-        llvm_unreachable("unsupported data flow pattern for invoke_simd 12");
+      if (!getHelperAndInvokeeIfInvokeSimdCall(CI).first) {
+        return false;
       }
       continue;
     }
-    if (const auto *LI = dyn_cast<LoadInst>(V)) {
+    if (isa<LoadInst>(V)) {
       // LoadInst from this addr is OK, as it does not affect what can be stored
       // through the addr
       continue;
     }
-    llvm_unreachable("unsupported data flow pattern for invoke_simd 13");
+    return false;
   }
+  return true;
 }
 
-// Example1 (function is direct argument to
-// _Z33__regcall3____builtin_invoke_simd):
-// %call6.i = call spir_func float @_Z33__regcall3____builtin_invoke_simd...(
-//   <16 x float> (float addrspace(4)*, <16 x float>, i32)* %28, <== function
-//   pointer float addrspace(4)* %arg1, float %arg2, i32 %arg3)
-//
-// Example 2 (invoke_simd's target function pointer flows through IR):
-// %fptr_t = <16 x float> (float addrspace(4)*, <16 x float>, i32)*
-// ...
-// %fa_as0 = alloca %fptr_t
-// ...
-// %fa = addrspacecast %fptr_t* %fa_as0 to %fptr_t addrspace(4)*
-// ...
-// store %fptr_t @__SIMD_CALLEE, %fptr_t addrspace(4)* %fa
-// ...
-// %f = load %fptr_t, %fptr_t addrspace(4)* %fa
-// ...
-// %res = call spir_func float @_Z33__regcall3____builtin_invoke_simd...(
-//   %fptr_t %f, <== function pointer
-//  float addrspace(4)* %arg1,
-//  float %arg2,
-//  i32 %arg3)
-//
-bool processInvokeSimdCall(CallInst *CI) {
-  Value *V = getInvokeeIfInvokeSimdCall(CI);
+// Deduce a single function whose address this value can only contain and
+// return it, otherwise (if can't be deduced or multiple functions deduced)
+// return nullptr.
+Function *deduceFunction(Value *I) {
+  while (true) {
+    I = stripCasts(I);
+    Function *Res = dyn_cast<Function>(I);
+
+    if (Res) {
+      return Res;
+    }
+    if (Argument *Arg = dyn_cast<Argument>(I)) {
+      // invoke_simd target is a function pointer which came via a formal
+      // parameter - do inter-procedural analysis trying to determine
+      // actual target
+      Function *Parent = Arg->getParent();
+      // follow all calls to F and see what functions were passed as actual
+      // argument
+      for (const User *U : Parent->users()) {
+        const auto *CI = dyn_cast<CallInst>(U);
+
+        if (!CI || (CI->getCalledFunction() != Parent)) {
+          llvm_unreachable("unsupported data flow pattern for invoke_simd A");
+        }
+        Value *ActualArg = CI->getArgOperand(Arg->getArgNo());
+        Function *F = deduceFunction(ActualArg);
```

Comment:
There is not protection from endless recursion.
It most likely not needed here because FrontEnd protects from cases like (A calls B, B calls A). Please confirm.

Reply:
Yes, that was my thought. SYCL does not support recursion yet. But I will actually add it to be on the safe side.